myapp/app/controllers/tasks_controller.rb

```diff
@@ -2,7 +2,9 @@ class TasksController < ApplicationController
   before_action :set_task, only: [:show, :edit, :update, :destroy]
 
   def index
-    @tasks = Task.all
+    sort_column = Task.column_names.include?(task_list_params[:sort_column]) ? task_list_params[:sort_column] : "created_at"
+    sort_direction = %w[ASC DESC].include?(task_list_params[:sort_direction]) ? task_list_params[:sort_direction] : "ASC"
+    @tasks = Task.all.order("#{sort_column} #{sort_direction}")
```

Comment:
It need to prevent SQL injection by passing the string received from the user directly!
Please think about  how to prevent SQL injection.

Reply:
https://rails-sqli.org/

Reply:
Thanks for telling me such a nice site about SQL injection in rails project! I’ve  gained a deeper understanding of the issue and learnt the importance of not passing strings from users directly.

And for the codes here, I think the params will be checked first to see if they’re included in the given range(if sort_column in the array of the table column names & if sort_direction in [ASC, DESC]). If not, a default value(created_at & ASC) will replace that. So, I think those malicious params(SQL injection codes) will be replaced to the default value here.
Maybe the problem is that I should change the code in line 7 to the following one to avoid using #{}?
`@tasks = Task.all.order(sort_column+ " " + sort_direction)`

I’m not clear which part will cause SQL injection here. Could you please kindly point out the part which may contain potential problems?

Reply:
Sorry:pray:, I missed the sanitization here!

```ruby
sort_column = Task.column_names.include?(task_list_params[:sort_column]) ? task_list_params[:sort_column] : "created_at"
sort_direction = %w[ASC DESC].include?(task_list_params[:sort_direction]) ? task_list_params[:sort_direction] : "ASC"
```

It would be better to implement it in `TasksController#task_list_params` for easier reuse.

Reply:
https://github.com/Fablic/training/pull/1564/commits/9273e6a9824fc0d81f4542f74c6982367714f374